### PR TITLE
Added debug backtrace info to json dbtj

### PR DESCRIPTION
**Pull request by fcasal**
Fixes https://github.com/radare/radare2/issues/10227

### CODE REVIEW

File: libr/core/cmd_debug.c [hunk 1]
```diff
@@ -2714,6 +2714,81 @@ static void asciiart_backtrace(RCore *core, RList *frames) {
...
+	if (!strcmp (flagdesc, flagdesc2)) {
+		flagdesc2[0] = 0;
+	}
+
+	if (hex_format) {
+		if (core->dbg->bits & R_SYS_BITS_64) {
+			snprintf (pcstr, sizeof_str, "0x%-16" PFMT64x, frame->addr);
+			snprintf (spstr, sizeof_str, "0x%-16" PFMT64x, frame->sp);
+		} else if (core->dbg->bits & R_SYS_BITS_32) {
+			snprintf (pcstr, sizeof_str, "0x%-8" PFMT64x, frame->addr);
```
Comment (radare) [inline libr/core/cmd_debug.c]: any reason for `%-` ?

File: libr/core/cmd_debug.c [hunk 2]
```diff
@@ -2714,6 +2714,81 @@ static void asciiart_backtrace(RCore *core, RList *frames) {
...
+						"%s+%d", f->name, delta + 1);
+			} else if (delta<0) {
+				snprintf (flagdesc2, sizeof_flags,
+						"%s%d", f->name, delta + 1);
+			} else {
+				snprintf (flagdesc2, sizeof_flags,
+						"%s+1", f->name);
+			}
+		} else {
+			snprintf (flagdesc2, sizeof_flags,
```
Comment (radare) [inline libr/core/cmd_debug.c]: this fits in 1 line, as well as all the other lines

File: libr/core/cmd_debug.c [hunk 3]
```diff
@@ -2714,6 +2714,81 @@ static void asciiart_backtrace(RCore *core, RList *frames) {
 	}
 }
 
+
+static void get_backtrace_info(RCore* core, RDebugFrame* frame, ut64 addr, char * flagdesc, char * flagdesc2, char *pcstr, char* spstr, bool hex_format) {
+	int sizeof_flags = 1024;
```
Comment (radare) [inline libr/core/cmd_debug.c]: ugly hack, asuming size of args is wrong

File: libr/core/cmd_debug.c [hunk 4]
```diff
@@ -2889,65 +2973,10 @@ static void r_core_cmd_bp(RCore *core, const char *input) {
 			}
 			i = 0;
 			list = r_debug_frames (core->dbg, addr);
+			bool hex_format = 1;
```
Comment (radare) [inline libr/core/cmd_debug.c]: if its bool use true or false, not 1

File: libr/core/cmd_debug.c [hunk 5]
```diff
@@ -2714,6 +2714,81 @@ static void asciiart_backtrace(RCore *core, RList *frames) {
...
 }
 
+
+static void get_backtrace_info(RCore* core, RDebugFrame* frame, ut64 addr, char * flagdesc, char * flagdesc2, char *pcstr, char* spstr, bool hex_format) {
+	int sizeof_flags = 1024;
+	int sizeof_str = 32;
+	RFlagItem *f = r_flag_get_at (core->flags, frame->addr, true);
+	flagdesc[0] = flagdesc2[0] = 0;
+	if (f) {
+		if (f->offset != addr) {
```
[the remark this answers is not in the document]
Comment (fcasal) [reply libr/core/cmd_debug.c]: I just copy'd the code from the switch to a function, I'm not sure how it works

File: libr/core/cmd_debug.c (hunk 1 above)
Comment (fcasal) [reply libr/core/cmd_debug.c]: I copied it from the code, but I guess its to align the addresses in the print

**Comment by radare** [comment]
no. the function that builds the strings should get char** and use r_str_newf instead of snprintf+stackvars

**Comment by radare** [comment]
Thanks!

**Comment by mcd1992** [comment]
@fcasal Seem to be getting a SIGSEGV related to this, I can reproduce it but only with a pre-existing project with a video game binary (can't redistribute).

get_backtrace_info frame: https://i.imgur.com/Eq3D4sD.jpg
__strcmp_ssse3 frame: https://i.imgur.com/medjSda.png

```
gef➤  bt
#0  0x00007fc404fc3df6 in __strcmp_ssse3 () from /usr/lib/libc.so.6
#1  0x00007fc40975308d in get_backtrace_info (core=0x55ccb4338560 <r>, frame=0x55ccb4c3eb60, addr=0xffffffffffffffff, flagdesc=0x7ffef2cacb28, flagdesc2=0x7ffef2cacb30, pcstr=0x7ffef2cacb38, spstr=0x7ffef2cacb40, hex_format=0x1) at cmd_debug.c:2752
#2  0x00007fc4097541eb in r_core_cmd_bp (core=0x55ccb4338560 <r>, input=0x55ccb4cf53b1 "bt") at cmd_debug.c:2973
#3  0x00007fc409759be2 in cmd_debug (data=0x55ccb4338560 <r>, input=0x55ccb4cf53b1 "bt") at cmd_debug.c:4485
#4  0x00007fc409812882 in r_cmd_call (cmd=0x55ccb47dc0a0, input=0x55ccb4cf53b0 "dbt") at cmd_api.c:237
#5  0x00007fc4097c9c7d in r_core_cmd_subst_i (core=0x55ccb4338560 <r>, cmd=0x55ccb4cf53b0 "dbt", colon=0x0, tmpseek=0x7ffef2cae176) at cmd.c:2700
#6  0x00007fc4097c6904 in r_core_cmd_subst (core=0x55ccb4338560 <r>, cmd=0x55ccb4cf53b0 "dbt") at cmd.c:1753
#7  0x00007fc4097cc3c7 in r_core_cmd (core=0x55ccb4338560 <r>, cstr=0x55ccb4860290 "dbt", log=0x1) at cmd.c:3402
#8  0x00007fc4097390f7 in r_core_prompt_exec (r=0x55ccb4338560 <r>) at core.c:2137
#9  0x000055ccb4135530 in main (argc=0x3, argv=0x7ffef2cae678, envp=0x7ffef2cae698) at radare2.c:1373
```

I can give a coredump via PM or something if needed. The issue seems to be `flagdesc` and `flagdesc2` are being set to "" and "h\r" instead of a char* to that string somewhere. Then the strcmp on line 2752 tries to deference \x00 in RDI. Gotta catch some :zzz:'s but I'll dig into it tomorrow some more if nothing comes to mind that would be causing it, haven't fully read what all this code does. Thanks.

**Comment by radare** [comment]
cant reproduce the crash but the json have too many spaces around :

[{"idx" : 0, "pc" : 4295012764, "sp" : 0, "frame_size" : 0,"fname" : "", "desc" : "ripmap.ls.r+20892"}]

**Comment by fcasal** [comment]
It crashes if you ood; dbt

**Comment by radare** [comment]
after my fix ?

**Comment by mcd1992** [comment]
Ah yea forgot to mention what commands I was doing `Po project; o filename; doo; dc; dbt`. Looks like it was fixed in b701e33ef

Thanks!

**Pull request merged**